library/std/src/panic.rs

```diff
@@ -202,5 +203,99 @@ pub fn always_abort() {
     crate::panicking::panic_count::set_always_abort();
 }
 
+/// The configuration for whether and how the default panic hook will capture
+/// and display the backtrace.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[unstable(feature = "panic_backtrace_config", issue = "93346")]
+#[non_exhaustive]
+pub enum BacktraceStyle {
+    /// Prints a terser backtrace which ideally only contains relevant
+    /// information.
+    Short,
+    /// Prints a backtrace with all possible information.
+    Full,
+    /// Does not capture or print a backtrace.
+    Disabled,
+    /// Backtraces aren't supported on this platform.
+    Unsupported,
+}
+
+impl BacktraceStyle {
+    pub(crate) fn full() -> Self {
+        if cfg!(feature = "backtrace") { BacktraceStyle::Full } else { BacktraceStyle::Disabled }
+    }
+
+    fn as_usize(self) -> usize {
+        match self {
+            BacktraceStyle::Short => 1,
+            BacktraceStyle::Full => 2,
+            BacktraceStyle::Disabled => 3,
+            BacktraceStyle::Unsupported => 4,
+        }
+    }
+}
+
+// Tracks whether we should/can capture a backtrace, and how we should display
+// that backtrace.
+//
+// Internally stores equivalent of an Option<BacktraceStyle>.
+static SHOULD_CAPTURE: AtomicUsize = AtomicUsize::new(0);
+
+/// Configure whether the default panic hook will capture and display a
+/// backtrace.
+///
+/// This is also configurable via the environment variable `RUST_BACKTRACE`, but
+/// usage of this function should be preferred from within Rust code.
+#[unstable(feature = "panic_backtrace_config", issue = "93346")]
+pub fn set_backtrace_style(style: BacktraceStyle) {
+    SHOULD_CAPTURE.store(style.as_usize(), Ordering::Release);
```

Comment:
This doesn't handle the case where the user passes in `Unsupported` as an input.

I doubt the utility of `Unsupported`. I suggest instead that on platforms for which backtraces are unsupported, this setting is just ignored, and no attempt be made to report `Unsupported` if/when queried.

Alternatively, perhaps a different enum should be used for setting vs. querying.

Reply:
Alternatively, collapse `Unsupported` and `Disabled`? Not sure what the practical difference is between the two.

Also, what is supposed to happen if the user calls `set_backtrace_style` to a style that this platform doesn't support? Should `set_backtrace_style` automatically map it so that `get_backtrace_style()` always returns the effective value?

Reply:
The minimal difference is that if it's not supported we won't suggest passing RUST_BACKTRACE=1 if it wasn't passed by the user. I'm not sure that tiny difference is worth the differentiation, though.

Good catch on not handling Unsupported as the input -- it's not really clear what to do in that case, I agree. Maybe panicking is appropriate. `get_backtrace_style` won't use the global static at all on platforms where it's not supported, so that variant of the enum isn't ever stored by the code there.

Reply:
I don't love the panicking option, how do you feel about just removing that variant and wrapping the style in an option, so that None is equivalent to the unsupported variant.

Reply:
I also would prefer to avoid panicking.

Reply:
Given brief discussion on the tracking issue I think I'll drop the getter for now, since I'm not a fan of the Option there, and we can just permit Short/Full/Off on the setter, mapping to the appropriate values. We should be able to skip doing anything if backtraces aren't supported; internally something closer to the current enum can be used.

Reply:
> Given brief discussion on the tracking issue I think I'll drop the getter for now,

Am I misunderstanding what you mean here? I expected to see that you removed the `get_backtrace_style()` function or marked it private, but it's still there.

Reply:
My original rewrite had an internal enum instead of the Option-wrapped external enum, but then once I moved to the Option it made things simpler and exposing it for the time being seems OK. I'll add an unresolved question/note that we may not want to stabilize the getter, though.